// @parcel/core/lib/resolveOptions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = resolveOptions;

function _utils() {
  const data = require("@parcel/utils");

  _utils = function () {
    return data;
  };

  return data;
}

function _loadDotEnv() {
  const data = _interopRequireDefault(require("./loadDotEnv"));

  _loadDotEnv = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _fs() {
  const data = require("@parcel/fs");

  _fs = function () {
    return data;
  };

  return data;
}

function _cache() {
  const data = _interopRequireDefault(require("@parcel/cache"));

  _cache = function () {
    return data;
  };

  return data;
}

function _packageManager() {
  const data = require("@parcel/package-manager");

  _packageManager = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Default cache directory name
const DEFAULT_CACHE_DIRNAME = '.parcel-cache';
const LOCK_FILE_NAMES = ['yarn.lock', 'package-lock.json', 'pnpm-lock.yaml']; // Generate a unique instanceId, will change on every run of parcel

function generateInstanceId(entries) {
  return (0, _utils().md5FromString)(`${entries.join(',')}-${Date.now()}-${Math.round(Math.random() * 100)}`);
}

async function resolveOptions(initialOptions) {
  var _initialOptions$mode, _initialOptions$minif, _initialOptions$publi, _initialOptions$shoul, _initialOptions$env, _initialOptions$shoul2, _initialOptions$hmrOp, _initialOptions$shoul3, _initialOptions$shoul4, _initialOptions$shoul5, _initialOptions$sourc, _initialOptions$scope, _initialOptions$logLe;

  let entries;

  if (initialOptions.entries == null || initialOptions.entries === '') {
    entries = [];
  } else if (Array.isArray(initialOptions.entries)) {
    entries = initialOptions.entries.map(entry => _path().default.resolve(entry));
  } else {
    entries = [_path().default.resolve(initialOptions.entries)];
  }

  let inputFS = initialOptions.inputFS || new (_fs().NodeFS)();
  let outputFS = initialOptions.outputFS || new (_fs().NodeFS)();
  let packageManager = initialOptions.packageManager || new (_packageManager().NodePackageManager)(inputFS);
  let entryRoot = initialOptions.entryRoot != null ? _path().default.resolve(initialOptions.entryRoot) : (0, _utils().getRootDir)(entries);

  let projectRootFile = (await (0, _utils().resolveConfig)(inputFS, _path().default.join(entryRoot, 'index'), [...LOCK_FILE_NAMES, '.git', '.hg'])) || _path().default.join(inputFS.cwd(), 'index'); // ? Should this just be rootDir


  let lockFile = null;

  let rootFileName = _path().default.basename(projectRootFile);

  if (LOCK_FILE_NAMES.includes(rootFileName)) {
    lockFile = projectRootFile;
  }

  let projectRoot = _path().default.dirname(projectRootFile);

  let inputCwd = inputFS.cwd();
  let outputCwd = outputFS.cwd();
  let cacheDir = // If a cacheDir is provided, resolve it relative to cwd. Otherwise,
  // use a default directory resolved relative to the project root.
  initialOptions.cacheDir != null ? _path().default.resolve(outputCwd, initialOptions.cacheDir) : _path().default.resolve(projectRoot, DEFAULT_CACHE_DIRNAME);
  let cache = new (_cache().default)(outputFS, cacheDir);
  let mode = (_initialOptions$mode = initialOptions.mode) !== null && _initialOptions$mode !== void 0 ? _initialOptions$mode : 'development';
  let minify = (_initialOptions$minif = initialOptions.minify) !== null && _initialOptions$minif !== void 0 ? _initialOptions$minif : mode === 'production';
  let publicUrl = (_initialOptions$publi = initialOptions.publicUrl) !== null && _initialOptions$publi !== void 0 ? _initialOptions$publi : '/';
  let distDir = initialOptions.distDir != null ? _path().default.resolve(inputCwd, initialOptions.distDir) : undefined;
  return {
    config: initialOptions.config,
    defaultConfig: initialOptions.defaultConfig,
    shouldPatchConsole: (_initialOptions$shoul = initialOptions.shouldPatchConsole) !== null && _initialOptions$shoul !== void 0 ? _initialOptions$shoul : process.env.NODE_ENV !== 'test',
    env: { ...process.env,
      ...initialOptions.env,
      ...(await (0, _loadDotEnv().default)((_initialOptions$env = initialOptions.env) !== null && _initialOptions$env !== void 0 ? _initialOptions$env : {}, inputFS, _path().default.join(projectRoot, 'index')))
    },
    mode,
    minify,
    shouldAutoInstall: (_initialOptions$shoul2 = initialOptions.shouldAutoInstall) !== null && _initialOptions$shoul2 !== void 0 ? _initialOptions$shoul2 : false,
    hmrOptions: (_initialOptions$hmrOp = initialOptions.hmrOptions) !== null && _initialOptions$hmrOp !== void 0 ? _initialOptions$hmrOp : null,
    shouldContentHash: (_initialOptions$shoul3 = initialOptions.shouldContentHash) !== null && _initialOptions$shoul3 !== void 0 ? _initialOptions$shoul3 : initialOptions.mode === 'production',
    serveOptions: initialOptions.serveOptions ? { ...initialOptions.serveOptions,
      distDir: distDir !== null && distDir !== void 0 ? distDir : _path().default.join(outputCwd, 'dist')
    } : false,
    shouldDisableCache: (_initialOptions$shoul4 = initialOptions.shouldDisableCache) !== null && _initialOptions$shoul4 !== void 0 ? _initialOptions$shoul4 : false,
    shouldProfile: (_initialOptions$shoul5 = initialOptions.shouldProfile) !== null && _initialOptions$shoul5 !== void 0 ? _initialOptions$shoul5 : false,
    cacheDir,
    entries,
    entryRoot,
    defaultEngines: initialOptions.defaultEngines,
    targets: initialOptions.targets,
    sourceMaps: (_initialOptions$sourc = initialOptions.sourceMaps) !== null && _initialOptions$sourc !== void 0 ? _initialOptions$sourc : true,
    scopeHoist: (_initialOptions$scope = initialOptions.scopeHoist) !== null && _initialOptions$scope !== void 0 ? _initialOptions$scope : initialOptions.mode === 'production',
    publicUrl,
    distDir,
    logLevel: (_initialOptions$logLe = initialOptions.logLevel) !== null && _initialOptions$logLe !== void 0 ? _initialOptions$logLe : 'info',
    projectRoot,
    lockFile,
    inputFS,
    outputFS,
    cache,
    packageManager,
    instanceId: generateInstanceId(entries),
    detailedReport: initialOptions.detailedReport
  };
}
